### PR TITLE
[SPARK-7990][SQL] Add methods to facilitate equi-join on multiple joining keys

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-7990

### CODE REVIEW

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -509,30 +509,42 @@ def join(self, other, joinExprs=None, joinType=None):
         The following performs a full outer join between ``df1`` and ``df2``.
 
         :param other: Right side of the join
-        :param joinExprs: a string for join column name, or a join expression (Column).
-            If joinExprs is a string indicating the name of the join column,
-            the column must exist on both sides, and this performs an inner equi-join.
+        :param joinExprs: a string for join column name, a list of column names,
```
Comment (rxin) [inline python/pyspark/sql/dataframe.py]: let's change the argument name in python as explained in the jira ticket

Comment (viirya) [inline python/pyspark/sql/dataframe.py]: ok.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 2]
```diff
@@ -448,6 +461,10 @@ class DataFrame private[sql](
    * @since 1.3.0
    */
   def join(right: DataFrame, joinExprs: Column, joinType: String): DataFrame = {
+    join(right, Seq(joinExprs), joinType)
+  }
+
+  def join(right: DataFrame, joinExprs: Seq[Column], joinType: String): DataFrame = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: i think we should remove this one for scala.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 3]
```diff
@@ -395,22 +395,35 @@ class DataFrame private[sql](
    * @since 1.4.0
    */
   def join(right: DataFrame, usingColumn: String): DataFrame = {
+    join(right, Seq(usingColumn))
+  }
+
+  def join(right: DataFrame, usingColumns: Seq[String]): DataFrame = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: add javadoc

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 4]
```diff
@@ -395,22 +395,35 @@ class DataFrame private[sql](
...
     val joined = sqlContext.executePlan(
       Join(logicalPlan, right.logicalPlan, joinType = Inner, None)).analyzed.asInstanceOf[Join]
 
-    // Project only one of the join column.
-    val joinedCol = joined.right.resolve(usingColumn)
+    // Project only one of the join columns.
+    val joinedCols = usingColumns.map(col => joined.right.resolve(col))
+    val condition = usingColumns.map { col =>
+      catalyst.expressions.EqualTo(joined.left.resolve(col), joined.right.resolve(col))
+    }.foldLeft[Option[catalyst.expressions.BinaryExpression]](None) { (opt, eqTo) =>
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: this can be simplifed into a reduceOption right?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Ah. yes. Did not notice that.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala (hunk 2 above)
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: So for `def join(self, other, joinExprs=None, joinType=None)` we don't support given a list of `Column` as `joinExprs`?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: for Python we can support it.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
@rxin Any more comments?

**Comment by rxin** [comment]
Thanks. I'm merging this.

**Pull request closed without merging**